lib/common/udp_tunnel.c

```diff
@@ -0,0 +1,106 @@
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+
+#include "h2o/httpclient.h"
+
+struct st_h2o_udp_tunnel_t {
+    h2o_httpclient_udp_tunnel_t super;
+    struct sockaddr_storage ss;
+    socklen_t len;
+    h2o_socket_t *sock;
+    h2o_loop_t *loop;
+};
+
+static void tunnel_on_destroy(h2o_httpclient_udp_tunnel_t *_tunnel)
+{
+    struct st_h2o_udp_tunnel_t *tunnel = (void *)_tunnel;
+
+    h2o_socket_close(tunnel->sock);
+    free(tunnel);
+}
+
+static void tunnel_socket_on_read(h2o_socket_t *sock, const char *err)
+{
+    struct st_h2o_udp_tunnel_t *tunnel = sock->data;
+
+    if (err != NULL) {
+        tunnel->super.on_read(&tunnel->super, err, NULL, 0);
+        return;
+    }
+
+    uint8_t buf[16384];
+    struct msghdr mess;
+    struct sockaddr sa;
+    struct iovec vec;
+    memset(&mess, 0, sizeof(mess));
+    mess.msg_name = &sa;
+    mess.msg_namelen = sizeof(sa);
+    vec.iov_base = buf;
+    vec.iov_len = sizeof(buf);
+    mess.msg_iov = &vec;
+    mess.msg_iovlen = 1;
+    ssize_t rret;
+    while ((rret = recvmsg(h2o_socket_get_fd(tunnel->sock), &mess, 0)) == -1 && errno == EINTR)
+        ;
+    tunnel->super.on_read(&tunnel->super, NULL, vec.iov_base, vec.iov_len);
+}
+
+
+static void tunnel_on_write(h2o_httpclient_udp_tunnel_t *_tunnel, const void *bytes, size_t len)
+{
+    struct st_h2o_udp_tunnel_t *tunnel = (void *)_tunnel;
+    int fd;
+    if (!tunnel->sock) {
+        struct sockaddr_in sin;
+        if ((fd = socket(PF_INET, SOCK_DGRAM, 0)) == -1) {
+            perror("failed to create UDP socket");
+            exit(EXIT_FAILURE);
+        }
+        memset(&sin, 0, sizeof(sin));
+        if (bind(fd, (void *)&sin, sizeof(sin)) != 0) {
+            perror("failed to bind bind UDP socket");
+            exit(EXIT_FAILURE);
+        }
```

Comment:
I'm not sure if we want to use an unconnected socket.

Unconnected socket has three notable properties:
* can receive from any address
* can send to any address
* send path is slower

For CONNECT-UDP, we do not need the first two properties, and therefore, it would make sense to use a connect socket to avoid the third property.

Separately, I think we can move the creation of the socket to the open function? I'm not sure if there's benefit in delaying the operation.